hydrolib/core/io/friction/models.py

```diff
@@ -124,15 +119,59 @@ class Comments(INIBasedModel.Comments):
         delimiter=" ",
     )
 
-    _make_lists = make_list_validator("frictionvalues")
+    @validator("levels", always=True)
+    @classmethod
+    def _validate_levels(cls, v, values):
+        foo = "bar"
+        if v is not None and (
+            values["numlevels"] is None or len(v) != values["numlevels"]
+        ):
+            raise ValueError(
+                f"Number of values for levels should be equal to the numLevels value (branchId={values.get('branchid', '')})."
+            )
+
+        return v
+
+    @validator("chainage", always=True)
+    @classmethod
+    def _validate_chainage(cls, v, values):
+        if v is not None and len(v) != values["numlocations"]:
+            raise ValueError(
+                f"Number of values for chainage should be equal to the numLocations value (branchId={values.get('branchid', '')})."
+            )
+
+        return v
+
+    @validator("frictionvalues", always=True)
+    @classmethod
+    def _validate_frictionvalues(cls, v, values):
+        # number of values should be equal to numlocations*numlevels
+        numvals = max(1, values["numlocations"]) * (
+            1
+            if (
+                "numlevels" not in values
+                or values["numlevels"] is None
+                or values["numlevels"] is 0
+            )
+            else values["numlevels"]
+        )
+        if v is not None and len(v) != numvals:
+            raise ValueError(
+                f"Number of values for frictionValues should be equal to the numLocations*numLevels value (branchId={values.get('branchid', '')})."
+            )
+
+        return v
 
 
 class FrictionModel(INIModel):
     general: FrictGeneral = FrictGeneral()
-    global_: FrictGlobal = Field(alias="global")  # to circumvent built-in kw
+    global_: List[FrictGlobal] = Field([], alias="global")  # to circumvent built-in kw
     branch: List[FrictBranch] = []
 
     _split_to_list = make_list_validator(
+        "global_",
+    )
+    _split_to_list2 = make_list_validator(
```

Comment:
Why was this changed?

Reply:
The file format in https://content.oss.deltares.nl/delft3d/manuals/D-Flow_FM_User_Manual_1D2D.pdf#section.C.15 defines:
> Multiple roughness variable definitions that don’t vary spatially (i.e. that don’t require Branch blocks) may be combined in one file. In that case there will one General block, multiple instances of the Global block and no Branch blocks in the file. See the second example.

So, our ```FrictionModel.global_``` must be a list, and not a scalar.

Reply:
Sorry, I was unclear, what I meant was: you don't need to call make_list_validator twice you can just do the following:

```
    _split_to_list = make_list_validator(
        "global_",
        "branch",
    )
```